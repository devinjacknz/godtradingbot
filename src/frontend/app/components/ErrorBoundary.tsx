'use client';

import { Component } from 'react';
import type { ReactNode, ErrorInfo } from 'react';
import { Box, Typography, Button, CircularProgress, Paper } from '@mui/material';
import { useRouter } from 'next/navigation';

interface Props {
  children: ReactNode;
}

interface State {
  hasError: boolean;
  error: Error | null;
  isRetrying: boolean;
}

class ErrorBoundaryBase extends Component<Props & { router: ReturnType<typeof useRouter> }, State> {
  constructor(props: Props & { router: ReturnType<typeof useRouter> }) {
    super(props);
    this.state = { hasError: false, error: null, isRetrying: false };
  }

  static getDerivedStateFromError(error: Error): State {
    return { hasError: true, error, isRetrying: false };
  }

  componentDidCatch(error: Error, info: ErrorInfo): void {
    console.error('ErrorBoundary caught an error:', error, info);
  }

  handleRetry = async (): Promise<void> => {
    this.setState({ isRetrying: true });
    try {
      await new Promise<void>(resolve => setTimeout(resolve, 2000));
      this.props.router.refresh();
      this.setState({ hasError: false, error: null, isRetrying: false });
    } catch (error) {
      this.setState({
        isRetrying: false,
        error: error instanceof Error ? error : new Error(String(error))
      });
    }
  };

  handleHome = async (): Promise<void> => {
    try {
      await this.props.router.push('/');
    } catch (error) {
      console.error('Navigation error:', error);
      window.location.href = '/';
    }
  };

  render() {
    if (!this.state.hasError) {
      return this.props.children;
    }

    const { error, isRetrying } = this.state;
    const errorMessage = error?.name === 'NetworkError' 
      ? 'Network error occurred'
      : error?.message || 'An unexpected error occurred';

    return (
      <Paper sx={{ p: 4, m: 2, bgcolor: '#1e1e1e' }} data-testid="error-boundary">
        <Box sx={{ textAlign: 'center' }}>
          <Typography variant="h5" color="error" gutterBottom>
            Something went wrong
          </Typography>
          <Typography color="text.secondary" sx={{ mb: 2 }} data-testid="error-message">
            {errorMessage}
          </Typography>
          {error?.stack && (
            <Typography 
              variant="body2" 
              sx={{ mb: 3, color: 'text.secondary', textAlign: 'left', whiteSpace: 'pre-wrap' }}
              data-testid="error-stack"
            >
              {error.stack}
            </Typography>
          )}
          <Box sx={{ display: 'flex', gap: 2, justifyContent: 'center' }}>
            <Button
              variant="contained"
              onClick={this.handleRetry}
              sx={{ mr: 2 }}
              data-testid="error-boundary-retry"
              disabled={isRetrying}
            >
              {isRetrying ? (
                <CircularProgress size={24} color="inherit" />
              ) : (
                'Try Again'
              )}
            </Button>
            <Button
              variant="outlined"
              onClick={this.handleHome}
              data-testid="error-boundary-home"
            >
              Return Home
            </Button>
          </Box>
        </Box>
      </Paper>
    );
  }
}

const ErrorBoundary = (props: Props) => {
  const router = useRouter();
  return <ErrorBoundaryBase {...props} router={router} />;
};

export default ErrorBoundary;
