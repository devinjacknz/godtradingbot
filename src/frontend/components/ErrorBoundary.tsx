import React, { Component, ErrorInfo, ReactNode } from 'react';
import { useToast } from './ui/use-toast';

interface Props {
  children: ReactNode;
}

interface State {
  hasError: boolean;
}

export class ErrorBoundary extends Component<Props, State> {
  public state: State = {
    hasError: false,
  };

  public static getDerivedStateFromError(_: Error): State {
    return { hasError: true };
  }

  public componentDidCatch(error: Error, errorInfo: ErrorInfo) {
    console.error('Uncaught error:', error, errorInfo);
    
    const { toast } = useToast(); toast({
      variant: 'destructive',
      title: 'Application Error',
      description: 'An unexpected error occurred. Please try refreshing the page.',
    });
  }

  public render() {
    if (this.state.hasError) {
      return (
        <div className="flex flex-col items-center justify-center min-h-screen p-4 space-y-4 text-center">
          <h1 className="text-2xl font-bold">Something went wrong</h1>
          <p className="text-muted-foreground">
            An unexpected error occurred. Please try refreshing the page.
          </p>
          <button
            className="px-4 py-2 text-sm font-medium text-white bg-primary rounded-md hover:bg-primary/90"
            onClick={() => window.location.reload()}
          >
            Refresh Page
          </button>
        </div>
      );
    }

    return this.props.children;
  }
}
