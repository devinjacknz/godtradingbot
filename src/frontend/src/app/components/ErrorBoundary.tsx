'use client';

import React, { Component, ErrorInfo, ReactNode } from 'react';
import { Alert, Button, Card, Typography, Box } from '@mui/material';
import { ErrorOutline, Refresh } from '@mui/icons-material';
import { useLanguage } from '../contexts/LanguageContext';
import { useDebug } from '../contexts/DebugContext';

interface Props {
  children: ReactNode;
  fallback?: ReactNode;
  onReset?: () => void;
  onError?: (error: Error, errorInfo: ErrorInfo) => void;
}

interface State {
  hasError: boolean;
  error: Error | null;
  errorInfo: ErrorInfo | null;
}

class ErrorBoundaryClass extends Component<Props, State> {
  public state: State = {
    hasError: false,
    error: null,
    errorInfo: null
  };

  public static getDerivedStateFromError(error: Error): State {
    return {
      hasError: true,
      error,
      errorInfo: null
    };
  }

  public componentDidCatch(error: Error, errorInfo: ErrorInfo) {
    this.setState({
      error,
      errorInfo
    });

    // 调用错误处理回调
    if (this.props.onError) {
      this.props.onError(error, errorInfo);
    }
  }

  private handleReset = () => {
    this.setState({
      hasError: false,
      error: null,
      errorInfo: null
    });

    // 调用重置回调
    if (this.props.onReset) {
      this.props.onReset();
    }
  };

  public render() {
    const { hasError, error, errorInfo } = this.state;
    const { children, fallback } = this.props;

    if (hasError) {
      // 如果提供了自定义fallback,则使用它
      if (fallback) {
        return fallback;
      }

      // 默认错误UI
      return (
        <ErrorDisplay
          error={error}
          errorInfo={errorInfo}
          onReset={this.handleReset}
        />
      );
    }

    return children;
  }
}

interface ErrorDisplayProps {
  error: Error | null;
  errorInfo: ErrorInfo | null;
  onReset: () => void;
}

const ErrorDisplay: React.FC<ErrorDisplayProps> = ({
  error,
  errorInfo,
  onReset
}) => {
  const { language } = useLanguage();
  const { isDebugMode, log } = useDebug();

  // 记录错误到调试日志
  React.useEffect(() => {
    if (error && isDebugMode) {
      log('error', 'ErrorBoundary', error.message, {
        error,
        errorInfo
      });
    }
  }, [error, errorInfo, isDebugMode, log]);

  return (
    <Box
      sx={{
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
        minHeight: '100vh',
        p: 3,
        bgcolor: 'background.default'
      }}
    >
      <Card
        sx={{
          maxWidth: 600,
          width: '100%',
          p: 4,
          textAlign: 'center'
        }}
      >
        <ErrorOutline
          color="error"
          sx={{ fontSize: 64, mb: 2 }}
        />

        <Typography variant="h5" gutterBottom color="error">
          {language === 'zh' ? '出错了!' : 'Something went wrong!'}
        </Typography>

        <Alert severity="error" sx={{ mb: 3 }}>
          {error?.message || (language === 'zh' ? '发生未知错误' : 'An unknown error occurred')}
        </Alert>

        {isDebugMode && errorInfo && (
          <Box
            sx={{
              mt: 2,
              mb: 3,
              p: 2,
              bgcolor: 'grey.100',
              borderRadius: 1,
              overflow: 'auto',
              maxHeight: 200
            }}
          >
            <Typography
              variant="body2"
              component="pre"
              sx={{
                fontFamily: 'monospace',
                whiteSpace: 'pre-wrap',
                wordBreak: 'break-word',
                m: 0
              }}
            >
              {errorInfo.componentStack}
            </Typography>
          </Box>
        )}

        <Button
          variant="contained"
          startIcon={<Refresh />}
          onClick={onReset}
          sx={{ mt: 2 }}
        >
          {language === 'zh' ? '重试' : 'Try Again'}
        </Button>

        {isDebugMode && (
          <Typography
            variant="caption"
            display="block"
            sx={{ mt: 2, color: 'text.secondary' }}
          >
            {language === 'zh'
              ? '错误已记录到调试日志'
              : 'Error has been logged to debug console'}
          </Typography>
        )}
      </Card>
    </Box>
  );
};

// 高阶组件用于包装类组件
export function withErrorBoundary<P extends object>(
  WrappedComponent: React.ComponentType<P>,
  errorBoundaryProps?: Omit<Props, 'children'>
) {
  return function WithErrorBoundaryWrapper(props: P) {
    return (
      <ErrorBoundaryClass {...errorBoundaryProps}>
        <WrappedComponent {...props} />
      </ErrorBoundaryClass>
    );
  };
}

// 用于函数组件的Hook
export function useErrorBoundary() {
  const [hasError, setHasError] = React.useState(false);
  const [error, setError] = React.useState<Error | null>(null);

  const handleError = (error: Error) => {
    setError(error);
    setHasError(true);
  };

  const reset = () => {
    setError(null);
    setHasError(false);
  };

  return {
    hasError,
    error,
    handleError,
    reset
  };
}

export default ErrorBoundaryClass;
