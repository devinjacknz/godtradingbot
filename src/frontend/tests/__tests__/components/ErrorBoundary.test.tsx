import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import ErrorBoundary from '@/app/components/ErrorBoundary';
import '@testing-library/jest-dom';
import type { ReactElement } from 'react';
import { useRouter } from 'next/navigation';

// Mock Next.js navigation and window.location
const mockPush = jest.fn();
const mockRefresh = jest.fn();
const mockReplace = jest.fn();

const originalLocation = window.location;
beforeAll(() => {
  delete (window as any).location;
  window.location = { ...originalLocation, href: '' };
});

afterAll(() => {
  window.location = originalLocation;
});

jest.mock('next/navigation', () => ({
  useRouter: () => ({
    push: mockPush,
    refresh: mockRefresh,
    replace: mockReplace
  })
}));

function TestComponent(): ReactElement {
  return <div data-testid="test-content">Test Content</div>;
}

describe('ErrorBoundary', () => {
  const consoleError = jest.spyOn(console, 'error').mockImplementation(() => {});

  beforeEach(() => {
    jest.clearAllMocks();
  });

  afterAll(() => {
    consoleError.mockRestore();
  });

  it('renders children when no error occurs', () => {
    const { container } = render(
      <ErrorBoundary>
        <TestComponent />
      </ErrorBoundary>
    );
    expect(screen.getByTestId('test-content')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('displays error UI when error occurs', async () => {
    function ThrowError(): ReactElement {
      throw new Error('Test error');
    }

    render(
      <ErrorBoundary>
        <ThrowError />
      </ErrorBoundary>
    );

    await waitFor(() => {
      expect(screen.getByText(/Something went wrong/i)).toBeInTheDocument();
      expect(screen.getByTestId('error-boundary-retry')).toBeInTheDocument();
      expect(screen.getByTestId('error-boundary-home')).toBeInTheDocument();
      expect(screen.getByText(/Test error/)).toBeInTheDocument();
    });
  });

  it('handles retry action', async () => {
    const user = userEvent.setup();
    function ThrowError(): ReactElement {
      throw new Error('Test error');
    }

    render(
      <ErrorBoundary>
        <ThrowError />
      </ErrorBoundary>
    );

    const retryButton = screen.getByTestId('error-boundary-retry');
    await user.click(retryButton);

    expect(mockRefresh).toHaveBeenCalled();
  });

  it('handles home navigation', async () => {
    const user = userEvent.setup();
    function ThrowError(): ReactElement {
      throw new Error('Test error');
    }

    render(
      <ErrorBoundary>
        <ThrowError />
      </ErrorBoundary>
    );

    const homeButton = screen.getByTestId('error-boundary-home');
    await user.click(homeButton);

    expect(mockPush).toHaveBeenCalledWith('/');
  });

  it('displays loading state during retry and recovers', async () => {
    const user = userEvent.setup();
    function ThrowError(): ReactElement {
      throw new Error('Test error');
    }

    render(
      <ErrorBoundary>
        <ThrowError />
      </ErrorBoundary>
    );

    const retryButton = screen.getByTestId('error-boundary-retry');
    await user.click(retryButton);

    await waitFor(() => {
      expect(retryButton).toBeDisabled();
      expect(screen.getByRole('progressbar')).toBeInTheDocument();
    });

    await waitFor(() => {
      expect(mockRefresh).toHaveBeenCalled();
    }, { timeout: 3000 });
  });

  it('handles navigation errors and recovery', async () => {
    const user = userEvent.setup();
    const consoleSpy = jest.spyOn(console, 'error').mockImplementation(() => {});
    const locationSpy = jest.spyOn(window.location, 'href', 'set');
    mockPush.mockRejectedValueOnce(new Error('Navigation failed'));
    
    let shouldThrow = true;
    function RecoverableNavigation(): ReactElement {
      if (shouldThrow) {
        throw new Error('Navigation error');
      }
      return <div data-testid="recovered-nav">Navigation successful</div>;
    }

    const { rerender } = render(
      <ErrorBoundary>
        <RecoverableNavigation />
      </ErrorBoundary>
    );

    const homeButton = screen.getByTestId('error-boundary-home');
    shouldThrow = false;
    await user.click(homeButton);

    // Should attempt Next.js navigation first
    expect(mockPush).toHaveBeenCalledWith('/');

    // Should log navigation error and fallback to window.location
    expect(consoleSpy).toHaveBeenCalledWith('Navigation error:', expect.any(Error));
    expect(locationSpy).toHaveBeenCalledWith('/');

    // Rerender to simulate recovery
    rerender(
      <ErrorBoundary>
        <RecoverableNavigation />
      </ErrorBoundary>
    );

    // Should show recovered content
    await waitFor(() => {
      expect(screen.queryByTestId('error-boundary')).not.toBeInTheDocument();
      expect(screen.getByTestId('recovered-nav')).toBeInTheDocument();
    });

    consoleSpy.mockRestore();
    locationSpy.mockRestore();
  });

  it('handles configuration and environment validation errors in trading system', async () => {
    const user = userEvent.setup();
    const consoleSpy = jest.spyOn(console, 'error').mockImplementation(() => {});
    type ConfigState = 'environment' | 'dependencies' | 'resources' | 'recovered';
    let currentState: ConfigState = 'environment';
    let configData: Record<string, any> = {};
    let configAttempts = 0;
    
    function ConfigErrorComponent(): ReactElement {
      switch (currentState) {
        case 'environment':
          configAttempts++;
          configData.environment = {
            component: 'EnvironmentValidator',
            variables: {
              api_keys: {
                required: ['DEEPSEEK_API_KEY', 'BINANCE_API_KEY'],
                missing: ['DEEPSEEK_API_KEY'],
                status: 'incomplete'
              },
              endpoints: {
                required: ['OLLAMA_ENDPOINT', 'TRADING_API'],
                invalid: ['TRADING_API'],
                status: 'misconfigured'
              }
            },
            runtime: {
              node_version: {
                required: '18.x',
                current: '16.x',
                status: 'outdated'
              },
              python_version: {
                required: '3.12',
                current: '3.8',
                status: 'incompatible'
              }
            },
            network: {
              ports: {
                required: [3000, 8000, 11434],
                blocked: [11434],
                status: 'unavailable'
              },
              firewall: {
                required_rules: ['allow_trading_api'],
                missing_rules: ['allow_trading_api'],
                status: 'blocked'
              }
            },
            attempt: configAttempts
          };
          throw new Error('Config Error: Trading environment validation failed');
        case 'dependencies':
          configData.dependencies = {
            component: 'DependencyValidator',
            packages: {
              npm: {
                missing: ['@mui/material', '@thirdweb-dev/react'],
                conflicts: ['react-router-dom'],
                status: 'incomplete'
              },
              python: {
                missing: ['deepseek-ai', 'fastapi'],
                version_mismatch: ['pandas'],
                status: 'incompatible'
              }
            },
            services: {
              docker: {
                required: ['ollama', 'trading-api'],
                unhealthy: ['ollama'],
                status: 'degraded'
              },
              external: {
                required: ['binance-api', 'deepseek-api'],
                unreachable: ['deepseek-api'],
                status: 'disconnected'
              }
            },
            system: {
              memory: {
                required: '4GB',
                available: '2GB',
                status: 'insufficient'
              },
              disk: {
                required: '10GB',
                available: '5GB',
                status: 'low'
              }
            }
          };
          throw new Error('Config Error: Trading dependencies validation failed');
        case 'resources':
          configData.resources = {
            component: 'ResourceManager',
            allocation: {
              cpu: {
                required: '4 cores',
                available: '2 cores',
                status: 'insufficient'
              },
              gpu: {
                required: 'cuda',
                available: 'none',
                status: 'missing'
              }
            },
            limits: {
              memory: {
                heap: {
                  maximum: '2GB',
                  used: '1.8GB',
                  status: 'critical'
                },
                cache: {
                  maximum: '1GB',
                  used: '900MB',
                  status: 'warning'
                }
              },
              storage: {
                temp: {
                  maximum: '5GB',
                  used: '4.5GB',
                  status: 'warning'
                },
                logs: {
                  maximum: '2GB',
                  used: '1.9GB',
                  status: 'critical'
                }
              }
            }
          };
          throw new Error('Config Error: Trading resources validation failed');
        case 'recovered':
          return <div data-testid="recovered-content">Trading configuration restored</div>;
      }
    }

    const { unmount, rerender } = render(
      <ErrorBoundary>
        <ConfigErrorComponent />
      </ErrorBoundary>
    );

    // Initial environment error state
    expect(screen.getByTestId('error-boundary')).toBeInTheDocument();
    expect(screen.getByText('Config Error: Trading environment validation failed')).toBeInTheDocument();
    expect(configAttempts).toBe(1);
    expect(configData.environment.variables.api_keys.status).toBe('incomplete');

    // Environment → Dependencies transition
    const retryButton = screen.getByTestId('error-boundary-retry');
    currentState = 'dependencies';
    await user.click(retryButton);

    await waitFor(() => {
      expect(retryButton).toBeDisabled();
      expect(screen.getByRole('progressbar')).toBeInTheDocument();
    });

    rerender(
      <ErrorBoundary>
        <ConfigErrorComponent />
      </ErrorBoundary>
    );

    // Verify dependencies error state
    await waitFor(() => {
      expect(screen.getByTestId('error-boundary')).toBeInTheDocument();
      expect(screen.getByText('Config Error: Trading dependencies validation failed')).toBeInTheDocument();
      expect(configData.dependencies.packages.npm.status).toBe('incomplete');
    });

    // Dependencies → Resources transition
    currentState = 'resources';
    await user.click(retryButton);

    rerender(
      <ErrorBoundary>
        <ConfigErrorComponent />
      </ErrorBoundary>
    );

    // Verify resources error state
    await waitFor(() => {
      expect(screen.getByTestId('error-boundary')).toBeInTheDocument();
      expect(screen.getByText('Config Error: Trading resources validation failed')).toBeInTheDocument();
      expect(configData.resources.allocation.cpu.status).toBe('insufficient');
    });

    // Resources → Recovered transition
    configData = {};
    currentState = 'recovered';
    await user.click(retryButton);

    rerender(
      <ErrorBoundary>
        <ConfigErrorComponent />
      </ErrorBoundary>
    );

    // Verify recovered state
    await waitFor(() => {
      expect(screen.queryByTestId('error-boundary')).not.toBeInTheDocument();
      expect(screen.getByTestId('recovered-content')).toBeInTheDocument();
      expect(screen.getByText('Trading configuration restored')).toBeInTheDocument();
      expect(Object.keys(configData)).toHaveLength(0);
    });

    // Test cleanup
    unmount();
    expect(screen.queryByTestId('error-boundary')).not.toBeInTheDocument();
    expect(screen.queryByTestId('recovered-content')).not.toBeInTheDocument();
    consoleSpy.mockRestore();
  });
});
