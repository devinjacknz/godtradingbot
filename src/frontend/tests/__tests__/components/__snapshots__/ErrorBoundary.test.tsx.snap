// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary renders children when no error occurs 1`] = `
<div>
  <div
    data-testid="test-content"
  >
    Test Content
  </div>
</div>
`;
